modules/distributor/forwarder.go

```diff
@@ -0,0 +1,317 @@
+package distributor
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/services"
+	"github.com/grafana/tempo/modules/overrides"
+	"github.com/grafana/tempo/pkg/util/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"go.uber.org/atomic"
+	"go.uber.org/multierr"
+)
+
+const (
+	defaultWorkerCount = 2
+	defaultQueueSize   = 100
+)
+
+var (
+	metricForwarderPushes = promauto.NewCounterVec(prometheus.CounterOpts{
+		Namespace: "tempo",
+		Name:      "distributor_forwarder_pushes_total",
+		Help:      "Total number of successful requests queued up for a tenant to the forwarder",
+	}, []string{"tenant"})
+	metricForwarderPushesFailures = promauto.NewCounterVec(prometheus.CounterOpts{
+		Namespace: "tempo",
+		Name:      "distributor_forwarder_pushes_failures_total",
+		Help:      "Total number of failed pushes to the queue for a tenant to the forwarder",
+	}, []string{"tenant"})
+	metricForwarderQueueLength = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: "tempo",
+		Name:      "distributor_forwarder_queue_length",
+		Help:      "Number of queued requests for a tenant",
+	}, []string{"tenant"})
+)
+
+type forwardFunc func(ctx context.Context, tenantID string, keys []uint32, traces []*rebatchedTrace) error
+
+type request struct {
+	keys   []uint32
+	traces []*rebatchedTrace
+}
+
+// forwarder queues up traces to be sent to the metrics-generators
+type forwarder struct {
+	services.Service
+
+	// per-tenant queue managers
+	queueManagers map[string]*queueManager
+	mutex         sync.RWMutex
+
+	forwardFunc forwardFunc
+
+	o                 *overrides.Overrides
+	overridesInterval time.Duration
+	shutdown          chan interface{}
+}
+
+func newForwarder(fn forwardFunc, o *overrides.Overrides) *forwarder {
+	rf := &forwarder{
+		queueManagers:     make(map[string]*queueManager),
+		mutex:             sync.RWMutex{},
+		forwardFunc:       fn,
+		o:                 o,
+		overridesInterval: time.Minute,
+		shutdown:          make(chan interface{}),
+	}
+
+	rf.Service = services.NewIdleService(rf.start, rf.stop)
+
+	return rf
+}
+
+// SendTraces queues up traces to be sent to the metrics-generators
+func (f *forwarder) SendTraces(ctx context.Context, tenantID string, keys []uint32, traces []*rebatchedTrace) {
+	select {
+	case <-f.shutdown:
+		return
+	default:
+	}
+
+	qm := f.getOrCreateQueueManager(tenantID)
+	if err := qm.pushToQueue(ctx, &request{keys: keys, traces: traces}); err != nil {
+		level.Error(log.Logger).Log("msg", "failed to push traces to queue", "tenant", tenantID, "err", err)
+		metricForwarderPushesFailures.WithLabelValues(tenantID).Inc()
+	} else {
+		metricForwarderPushes.WithLabelValues(tenantID).Inc()
+	}
```

Comment:
I think we should always update the total pushes metric. Looking at other places, we usually record the total amount of requests (success + failure) and the failures. This allows you to calculate %age of failures as `failed / total`.

For instance, when ingested spans we always increment `metricSpansIngested` and increment `metricDiscardedSpans` if the spans are dropped:

https://github.com/grafana/tempo/blob/a9af8352db92b78c8550ff72d3ede5112150266f/modules/distributor/distributor.go#L297-L308

So I think we should move the metric out of the else-case so it's updated for every request. (We'll also have to update the metric description.)

```suggestion
	}
	metricForwarderPushes.WithLabelValues(tenantID).Inc()
```

Reply:
Good point, changed